### PR TITLE
sql: txnState.sp is nil if COCKROACH_TRACE_SQL is used

### LINKED ISSUE #13837: tracing: the CDB process crash if enable COCKROACH_TRACE_SQL
The header (i.e. the first few lines) of the most recent log file. 
````
I170228 11:06:58.060465 1 util/log/clog.go:999  [config] file created at: 2017/02/28 11:06:58
I170228 11:06:58.060465 1 util/log/clog.go:999  [config] running on machine: gzns-dba-sz-ndbtest-02
I170228 11:06:58.060465 1 util/log/clog.go:999  [config] binary: CockroachDB CCL beta-20170223 (linux amd64, built 2017/02/23 16:30:53, go1.7.5)
I170228 11:06:58.060465 1 util/log/clog.go:999  [config] arguments: [/home/work/setup-cockroachdb/bin//cockroach start --port=26257 --http-port=8080 --store=path=/ssd/newdb,attrs=ssd --join=gzns-dba-sz-ndbtest-01.gzns --join=gzns-dba-sz-ndbtest-02.gzns --join=gzns-dba-sz-ndbtest-03.gzns --join=gzns-dba-sz-ndbtest-04.gzns --cache=24G --attrs=gzns --insecure]
````

Here is the call stack:
````
panic: No span in context? Was resetForNewSQLTxn() called previously? [recovered]
        panic: SELECT timestamp, eventType, targetID, reportingID, info, uniqueID FROM system.eventlog WHERE true ORDER BY timestamp DESC LIMIT $1: No span in context? Was resetForNewSQLTxn() called previously?
        panic: session: unexpected leftover memory: 12190 bytes
goroutine 3338 [running]:
panic(0x196c140, 0xc420ed2150)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/sql/mon.(*MemoryMonitor).Stop(0xc421f819a0, 0x7f4ff95844a8, 0xc422b860c0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/mon/mem_usage.go:339 +0x381
github.com/cockroachdb/cockroach/pkg/sql.(*Session).Finish(0xc421f80000, 0xc420444000)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:178 +0x99
panic(0x19ba540, 0xc420ed2120)
        /usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc421b34240, 0x83)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:484 +0x149
panic(0x196c140, 0xc420ed20e0)
        /usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql.(*txnState).finishSQLTxn(0xc421f80068, 0x7f4ff95844a8, 0xc422b860c0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:371 +0x26f
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc420444000, 0xc421f80000, 0xc421b34240, 0x83, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:742 +0xd45
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc420444000, 0xc421f80000, 0xc421b34240, 0x83, 0xc422cfc758, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:490 +0x11f
github.com/cockroachdb/cockroach/pkg/server.(*adminServer).Events(0xc4204f2000, 0x7f4ff95843c8, 0xc420d3ad20, 0xc420ed4720, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/admin.go:706 +0x473
````

### PR DESCRIPTION
**Pull request by petermattis**
Fixes #13837

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13868)

### CODE REVIEW

File: pkg/sql/session.go [hunk 1]
```diff
@@ -368,12 +370,12 @@ func (ts *txnState) resetStateAndTxn(state TxnStateEnum) {
 // The session context is just used for logging the SQL trace.
 func (ts *txnState) finishSQLTxn(sessionCtx context.Context) {
 	ts.mon.Stop(ts.Ctx)
-	if ts.sp == nil {
-		panic("No span in context? Was resetForNewSQLTxn() called previously?")
```
Comment (petermattis) [inline pkg/sql/session.go]: @andreimatei This panic implies this shouldn't happen, yet looking at `resetForNewSQLTxn` it seems that it always occurs if `traceSQL` is true. Am I missing something?

File: pkg/sql/session.go [hunk 2]
```diff
@@ -368,12 +370,12 @@ func (ts *txnState) resetStateAndTxn(state TxnStateEnum) {
...
+	var sampledFor7881 bool
+	if ts.sp != nil {
+		sampledFor7881 = (ts.sp.BaggageItem(keyFor7881Sample) != "")
```
Comment (tamird) [inline pkg/sql/session.go]: Unnecessary parens

**Review by tamird — APPROVED** [review]

File: pkg/sql/session.go (hunk 2 above)
Comment (petermattis) [inline pkg/sql/session.go]: For safety, right? The code feels less readable without them, though I don't particularly care either way, so I've left it alone.

**Comment by andreimatei** [comment]
I think your change was failing to close the span in the TRACE_SQL case. Let's do #13871, I think that's the right fix.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13868#-:-Ke9Y-XSBODnrSTZwXta:b-jt06qz)*

**Pull request closed without merging**